### PR TITLE
Add dml migrations

### PR DESCRIPTION
**Pull request by koletzilla**
## Summary
Related to https://github.com/ClickHouse/clickhouse-fivetran-destination/issues/85

Adds the 13 pending migrations defined in https://github.com/fivetran/fivetran_partner_sdk/blob/main/schema-migration-helper-service.md and tests.

The 4 related to live sync are not implemented as these are not supported by the partner-built adapters.

Structure:
- server_migrate.go: New gRPC service to cover the new calls. Follows the same schema as server.go, contain the gRPC management code.
- clickhouse_migrate.go: Same schema as clickhouse.go but to contain the cordination of the operations needed to be done for the new migrations.
- sql_migrate.go: Same schema as sql.go but to contain all the new sql needed for the new transformations

### CODE REVIEW

File: destination/db/clickhouse_migrate.go [hunk 1]
```diff
@@ -0,0 +1,636 @@
...
+	ctx context.Context,
+	schemaName string,
+	tableName string,
+	column string,
+	chType string,
+	comment string,
+	defaultValue string,
+) error {
+	// Step 1: Add column
+	addOp := &types.AlterTableOp{
```
Comment (BentsiLeviav) [inline destination/db/clickhouse_migrate.go:207]: A small question - for `ADD_COLUMN_WITH_DEFAULT_VALUE` - I noticed this in the migration guide https://github.com/fivetran/fivetran_partner_sdk/blob/56f52c3ed4f464390cf00a3a36dd9b5b9760cf73/schema-migration-helper-service.md?plain=1#L83 - Does it mean we'd better change the add column to something like `ALTER TABLE … ADD COLUMN IF NOT EXISTS`, or check for the column existence before?

If we do so, can we also add a test to cover this situation? (`ADD_COLUMN_WITH_DEFAULT_VALUE` when the column already exists)

File: destination/service/server_migrate.go [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+	op *pb.TableSyncModeMigrationOperation,
+) (*pb.MigrateResponse, error) {
+	migrationType := op.GetType()
+	log.Info(fmt.Sprintf("[Migrate] Sync mode migration %s for %s.%s", migrationType.String(), schema, table))
+
+	switch migrationType {
+	case pb.TableSyncModeMigrationType_SOFT_DELETE_TO_HISTORY:
+		softDeletedCol := op.GetSoftDeletedColumn()
+		if softDeletedCol == "" {
+			softDeletedCol = "_fivetran_deleted"
```
Comment (BentsiLeviav) [inline destination/service/server_migrate.go]: I see multiple usages of the string `_fivetran_deleted` - can you verify if we can use  `constants.FivetranDeleted` instead?

File: destination/service/server_migrate.go [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+	}
+	if !strings.HasPrefix(chType.Type, "Nullable(") {
+		chType.Type = fmt.Sprintf("Nullable(%s)", chType.Type)
+	}
+	return chType, nil
+}
+
+// formatMigrateValue formats a value for use in SQL UPDATE statements based on the Fivetran DataType.
+// For datetime types, converts ISO 8601 to nanosecond timestamp format.
+func formatMigrateValue(value string, colType pb.DataType) string {
```
Comment (BentsiLeviav) [inline destination/service/server_migrate.go]: Small style comment - `formatMigrateValue` and `parseTimestampToNanos` contain the same two-step parse pattern (time.RFC3339 → fallback format). Should it be extracted into a shared parseUTCTimestamp helper to avoid duplication?

File: destination/service/server_migrate.go [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+		return FailedMigrateResponse(schema, table, err), nil
+	}
+}
+
+func handleCopyOperation(
+	ctx context.Context,
+	conn *db.ClickHouseConnection,
+	schema string,
+	table string,
+	copy *pb.CopyOperation,
```
Comment (BentsiLeviav) [inline destination/service/server_migrate.go]: Minor comment - `copy` is a built-in function in Go, can we rename that one?

File: destination/db/clickhouse_migrate.go [hunk 5]
```diff
@@ -0,0 +1,636 @@
...
+	// Get column names for the INSERT...SELECT
+	tableDesc, err := conn.DescribeTable(ctx, schemaName, fromTable)
+	if err != nil {
+		return err
+	}
+	colNames := make([]string, len(tableDesc.Columns))
+	for i, col := range tableDesc.Columns {
+		colNames[i] = col.Name
+	}
+	insertStmt, err := sql.GetInsertFromSelectStatement(schemaName, fromTable, toTable, colNames)
```
Comment (BentsiLeviav) [inline destination/db/clickhouse_migrate.go]: Just to make sure it was intentional - I see we use in many place `FINAL` - but not here.

File: destination/db/clickhouse_migrate.go [hunk 6]
```diff
@@ -0,0 +1,636 @@
...
+	if err != nil {
+		waitErr := conn.WaitAllMutationsCompleted(ctx, err, schemaName, tableName)
+		if waitErr != nil {
+			return waitErr
+		}
+	}
+	return nil
+}
+
+func (conn *ClickHouseConnection) MigrateCopyTable(
```
Comment (BentsiLeviav) [inline destination/db/clickhouse_migrate.go:132]: A small question about the flow - if we create the table successfully, and then fail on the describe step or the subsequent insert-into-select fails, would the new empty `toTable` be left behind? Cause the next attempt might get "table already exists."

File: destination/db/sql/sql_migrate.go [hunk 7]
```diff
@@ -0,0 +1,487 @@
...
+
+	selectFromPart := fmt.Sprintf("SELECT %s FROM %s FINAL",
+		selectExprs,
+		fullTableName,
+	)
+
+	// Step 3: Build the WHERE part: WHERE _fivetran_active  AND _fivetran_start < {operation_timestamp}
+	// For DROP: only insert new versions for rows where the column actually has a value (AND {column_name} IS NOT NULL)
+
+	wherePart := fmt.Sprintf("WHERE `_fivetran_active` AND `_fivetran_start` < '%s'", operationTimestampNanos)
```
Comment (BentsiLeviav) [inline destination/db/sql/sql_migrate.go]: Minor style question - can we use `identifier(constants.FivetranActive)` and `identifier(constants.FivetranStart)` ?

File: destination/db/sql/sql_migrate.go [hunk 8]
```diff
@@ -0,0 +1,487 @@
...
+			pkColsList.WriteString(",")
+		}
+	}
+	pkCols := pkColsList.String()
+
+	deletedExpr := fmt.Sprintf("if(%s = true, false, true)", identifier(constants.FivetranActive))
+
+	var whereClause string
+	if !keepDeletedRows {
+		whereClause = fmt.Sprintf(" WHERE %s = true", identifier(constants.FivetranActive))
```
Comment (BentsiLeviav) [inline destination/db/sql/sql_migrate.go:361]: A minor style comment (feel free to ignore) - there are some places in this file where we check bools with 0/1 `"if(%s = 0, true, false)"` and some with true/false, worth alignment.

**Review by BentsiLeviav — COMMENTED** [review]
Also ran Claude to provide comments on tests, I beleive some of it is just noise, but still worth checking, LMK what you think:

---
  sql_migrate_test.go — Well covered overall

  What's well covered: All 9 SQL builder functions have happy-path tests and error-case tests. Special character escaping is tested (O'Brien, backslash). Both keepDeletedRows variants of
  GetInsertFromSelectHistoryToSoftDeleteStatement are tested. The same-timestamp composability case for GetInsertNewActiveVersionsStatement (with history columns in the input) is tested.

  Missing:

  - GetInsertNewActiveVersionsStatement — no test for empty operationTimestampNanos. The function validates it but there's no corresponding test case, unlike every other validated parameter in the file.
  - GetInsertNewActiveVersionsStatement — no test for empty overrideColumn. As noted in the analysis, this generates invalid SQL but there's no guard and no test.
  - GetInsertFromSelectHistoryToSoftDeleteStatement — only tested with a single PK column. No test with composite PKs, which affects the ORDER BY pk1, pk2, _fivetran_start DESC LIMIT 1 BY pk1, pk2 generation.
  - subtractOneMillisecond — only tested indirectly via GetCloseActiveRowsStatement. No direct test for the boundary case where the input nanosecond value is very small (underflow risk).

  ---
  server_migrate_test.go — Good coverage of edge cases, one important gap

  What's well covered: All unsupported/default entity branches, all timestamp parsing failures, all unknown column type failures, all four LIVE mode unsupported transitions, formatMigrateValue for all types including
  subsecond UTC_DATETIME.

  Missing:

  - handleUpdateColumnValue — no test for either NULL detection path. The AGENTS.md gotcha explicitly calls out that value == "" AND value == "NULL" both map to isNull = true. This is critical logic with no unit test:

  // Not tested:
  isNull := value == "" || value == "NULL"

  - handleTableSyncModeMigration — no test for SOFT_DELETE_TO_HISTORY or HISTORY_TO_SOFT_DELETE with an empty soft_deleted_column, verifying it falls back to _fivetran_deleted. Easy to add, documents a real behavioral
  contract.
  - migrateColumnType — no test for a type that is already Nullable(...). The guard if !strings.HasPrefix(chType.Type, "Nullable(") is untested. It's a defensive check that would double-wrap types without it.
  - The formatMigrateValue inconsistency (not called for AddColumnInHistoryMode) has no test demonstrating the divergence between the two add operations for UTC_DATETIME defaults. Adding a test that shows
  AddColumnInHistoryMode receives a raw ISO string while AddColumnWithDefaultValue receives nanoseconds would make the bug immediately visible.

  ---
  clickhouse_migrate_test.go — Too thin

  All seven tests are identical in structure: pass an empty schema name, expect an error. They only cover the SQL-builder-error propagation path and nothing else. More importantly, several functions have no test at all:

  - MigrateAddColumnInHistoryMode — no test, not even the SQL-builder-error path.
  - MigrateDropColumnInHistoryMode — no test.
  - MigrateSoftDeleteToHistory — no test.
  - MigrateHistoryToSoftDelete — no test.
  - MigrateCopyColumn — the "source column not found" error path (when fromColumn doesn't exist in the table) has no test.

  ---
  main_e2e_test.go — Solid but two important data gaps

  What's well covered: All 11 DML operations are exercised end-to-end. The same-timestamp composability case for add+drop history mode is covered. Schema (column list) is verified for every migrated table.

  Missing:

  - TestSchemaMigrationsDML verifies columns for transaction_renamed but not its data. Since transaction_renamed is the result of copy_table (the most data-sensitive operation in the DML test), its rows should be verified,
   not just its schema.
  - TestSchemaMigrationsDML verifies operation_time column exists but never checks its values. The whole point of add_column_with_default_value is that existing rows get the default. A query checking operation_time IS NOT
  NULL and matching the expected nanosecond value would verify the actual behavior.
  - TestSchemaMigrationsSyncModes tests HISTORY_TO_SOFT_DELETE but all source rows are active (_fivetran_active = true). There's no test where some rows are deleted (_fivetran_active = false), which is the main scenario
  this migration exists for — verifying those rows get _fivetran_deleted = true.
  - No e2e test for the validateHistoryModeTable empty-table skip path — running a history mode operation on an empty table and verifying it returns success without doing anything.

**Comment by koletzilla** [comment]
I have been working on the comment you shared:
- sql_migrate_test.go 
	- Added test mentioned. To make also the code cleaner, I have moved all logic and tests for empty values to the GRPC boundary (server_migate.go) so we can just validate it once there instead of having to check these values in internal calls.
- server_migrate_test.go
	- Last refactor fixed some of these issues like the NULL testing. Added tests for the missing soft_deleted_column
- clickhouse_migrate_test.go
	- Basically we are testing all important logic here with the e2e. The only remaining things to tests are to validate that the errors are correctly sent and seems like it may not be really worth it to at that many tests to cover these situations. In the coverage report we can see virtually all red lines are or `return err` or a `log.warn/error`
- main_e2e_test.go
  - Added tests for these steps

File: destination/db/clickhouse_migrate.go (hunk 1 above)
Comment (koletzilla) [reply destination/db/clickhouse_migrate.go:207]: I forgot that point :/ Added `IF NOT EXISTS` + a test

File: destination/service/server_migrate.go (hunk 3 above)
Comment (koletzilla) [reply destination/service/server_migrate.go]: Good point. Also the `time.RFC3339` wasn't correct, we only need to parse from the existing `constants.UTCDateTimeFormat` format. Removed that one and only kept one version to solve both cases

File: destination/service/server_migrate.go (hunk 2 above)
Comment (koletzilla) [reply destination/service/server_migrate.go]: I applied it in other of the latest commits. Now we should only have the string in tests :D

File: destination/service/server_migrate.go (hunk 4 above)
Comment (koletzilla) [reply destination/service/server_migrate.go]: Pushed a fix for that. Also I have found out that the linter can catch these errors, we should just activate it. There are a few other linters that are useful, so I have activated a bunch of them here: https://github.com/ClickHouse/clickhouse-fivetran-destination/pull/97

File: destination/db/sql/sql_migrate.go (hunk 7 above)
Comment (koletzilla) [reply destination/db/sql/sql_migrate.go]: fixed!

File: destination/db/sql/sql_migrate.go (hunk 8 above)
Comment (koletzilla) [reply destination/db/sql/sql_migrate.go:361]: fixed!

File: destination/db/clickhouse_migrate.go (hunk 5 above)
Comment (koletzilla) [reply destination/db/clickhouse_migrate.go]: Added final to this and to the query for GetMaxFivetranStartQuery that also required it

File: destination/db/clickhouse_migrate.go (hunk 6 above)
Comment (koletzilla) [reply destination/db/clickhouse_migrate.go:132]: Good one. I have added the same pattern we do in dbt: `drop target if exists` before recreating it again.

**Review by BentsiLeviav — APPROVED** [review]
